File: platform/src/main/java/com/toocol/termio/platform/nativefx/NativeBinding.java
package com.toocol.termio.platform.nativefx;

import com.toocol.termio.utilities.utils.OsUtil;
import javafx.scene.input.MouseEvent;
import javafx.scene.input.ScrollEvent;
import org.apache.commons.io.IOUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class NativeBinding {

    private static boolean initialized;

    private NativeBinding() {
        throw new AssertionError("Don't instantiate me!");
    }

    private static String libEnding() {
        if (isOS("windows")) return ".dll";
        if (isOS("linux")) return ".so";
        if (isOS("macos")) return ".dylib";
        return ".so";
    }

    private static String archName() {
        String osArch = System.getProperty("os.arch");

        if (osArch.toLowerCase().contains("x64")
                || osArch.toLowerCase().contains("amd64")
                || osArch.toLowerCase().contains("x86_64")) {
            return "x64";
        } else if (osArch.toLowerCase().contains("x86")) {
            return "x86";
        } else if (osArch.toLowerCase().contains("aarch64")) {
            return "aarch64";
        }

        return osArch;
    }

    private static String osName() {
        String osName = System.getProperty("os.name");

        if (osName.toLowerCase().contains("windows")) {
            return "windows";
        } else if (osName.toLowerCase().contains("linux")) {
            return "linux";
        } else if (osName.toLowerCase().contains("mac")
                || osName.toLowerCase().contains("darwin")) {
            return "macos";
        }

        return osName;
    }

    private static boolean isOS(String os) {
        return os.equals(osName());
    }

    public static void init() {
        if (initialized) {
            return;
        }

        initialized = true;

        String name = "nativenode" + OsUtil.libSuffix();
        InputStream inputStream;
        OutputStream outputStream;

        try {
            inputStream = NativeBinding.class.getResourceAsStream("/" + name);
            assert inputStream != null;

            String libraryPath = System.getenv("JAVA_HOME") + OsUtil.fileSeparator() + "bin" + OsUtil.fileSeparator();
            File fileOut = new File(libraryPath + name);
            outputStream = new FileOutputStream(fileOut);
            IOUtils.copy(inputStream, outputStream);

            String extractPath = fileOut.toString();

            inputStream.close();
            outputStream.close();

            System.load(extractPath);
        } catch (Exception e) {
            System.out.println("Load nativenode.dll failed. message = " + e);
            System.exit(-1);
        }
    }

    interface IntEnum {
        int asInt();

        static int combine(IntEnum... enums) {
            int result = 0;

            for (IntEnum e : enums) {
                result = result | e.asInt();
            }

            return result;
        }

        static int combine(int combined, IntEnum... enums) {
            int result = combined;

            for (IntEnum e : enums) {
                result = result | e.asInt();
            }

            return result;
        }
    }

    interface MouseBtn {
        int NO_BTN = 0;
        int PRIMARY_BTN = 1;
        int SECONDARY_BTN = 2;
        int MIDDLE_BTN = 4;

        static int fromEvent(MouseEvent ev) {
            int result = MouseBtn.NO_BTN;

            if (ev.isPrimaryButtonDown()) {
                result |= MouseBtn.PRIMARY_BTN;
            }

            if (ev.isSecondaryButtonDown()) {
                result |= MouseBtn.SECONDARY_BTN;
            }

            if (ev.isMiddleButtonDown()) {
                result |= MouseBtn.MIDDLE_BTN;
            }

            return result;
        }

        static int fromEvent(ScrollEvent ev) {
            // TODO implement me 8.07.2019
            return MouseBtn.NO_BTN;
        }
    }

    interface Modifier {
        int NO_KEY = 0;
        int SHIFT_KEY = 1;
        int ALT_KEY = 2;
        int META_KEY = 4;
        int CONTROL_KEY = 8;

        static int fromEvent(MouseEvent ev) {
            int result = Modifier.NO_KEY;

            if (ev.isShiftDown()) {
                result |= Modifier.SHIFT_KEY;
            }

            if (ev.isAltDown()) {
                result |= Modifier.ALT_KEY;
            }

            if (ev.isMetaDown()) {
                result |= Modifier.META_KEY;
            }

            if (ev.isControlDown()) {
                result |= Modifier.CONTROL_KEY;
            }

            return result;
        }

        static int fromEvent(ScrollEvent ev) {
            int result = Modifier.NO_KEY;

            if (ev.isShiftDown()) {
                result |= Modifier.SHIFT_KEY;
            }

            if (ev.isAltDown()) {
                result |= Modifier.ALT_KEY;
            }

            if (ev.isMetaDown()) {
                result |= Modifier.META_KEY;
            }

            if (ev.isControlDown()) {
                result |= Modifier.CONTROL_KEY;
            }

            return result;
        }
    }

    private static final Map<Integer, List<NativeEventListener>> listeners = new HashMap<>();

    public static void addEventListener(int key, NativeEventListener l) {
        // create list if not present
        if (!listeners.containsKey(key)) {
            List<NativeEventListener> list = new ArrayList<>();
            listeners.put(key, list);
        }

        listeners.get(key).add(l);
    }

    public static void removeEventListener(int key, NativeEventListener l) {
        // create list if not present
        if (!listeners.containsKey(key)) {
            List<NativeEventListener> list = new ArrayList<>();
            listeners.put(key, list);
        }

        listeners.get(key).remove(l);
    }

    public static void removeEventListeners(int key) {
        // create list if not present
        if (!listeners.containsKey(key)) {
            List<NativeEventListener> list = new ArrayList<>();
            listeners.put(key, list);
        }

        listeners.get(key).clear();
    }

    /*CALLED FROM NATIVE*/
    public static void fireNativeEvent(int key, String type, String evt) {
        // return early if not present
        if (!listeners.containsKey(key)) {
            return;
        }

        List<NativeEventListener> list = listeners.get(key);

        for (NativeEventListener l : list) {
            l.event(key, type, evt);
        }
    }

    public static native int nextKey();

    public static native int connectTo(String name);

    public static native boolean terminate(int key);

    public static native boolean isConnected(int key);

    public static native String sendMsg(int key, String msg);

    public static native void processNativeEvents(int key);

    public static native void resize(int key, int w, int h);

    public static native boolean isDirty(int key);

    public static native void redraw(int key, int x, int y, int w, int h);

    public static native void setDirty(int key, boolean value);

    public static native void setBufferReady(int key, boolean value);

    public static native boolean isBufferReady(int key);

    public static native int getW(int key);

    public static native int getH(int key);

    public static native boolean fireMousePressedEvent(int key, double x, double y, int buttons, int modifiers, long timestamp);

    public static native boolean fireMouseReleasedEvent(int key, double x, double y, int buttons, int modifiers, long timestamp);

    public static native boolean fireMouseClickedEvent(int key, double x, double y, int buttons, int modifiers, int clickCount, long timestamp);

    public static native boolean fireMouseEnteredEvent(int key, double x, double y, int buttons, int modifiers, long timestamp);

    public static native boolean fireMouseExitedEvent(int key, double x, double y, int buttons, int modifiers, long timestamp);

    public static native boolean fireMouseMoveEvent(int key, double x, double y, int buttons, int modifiers, long timestamp);

    public static native boolean fireMouseWheelEvent(int key, double x, double y, double amount, int buttons, int modifiers, long timestamp);

    public static native boolean fireKeyPressedEvent(int key, String characters, int keyCode, int modifiers, long timestamp);

    public static native boolean fireKeyReleasedEvent(int key, String characters, int keyCode, int modifiers, long timestamp);

    public static native boolean fireKeyTypedEvent(int key, String characters, int keyCode, int modifiers, long timestamp);

    public static native boolean requestFocus(int key, boolean focus, long timestamp);

    public static native ByteBuffer getBuffer(int key);

    public static native boolean lock(int key);

    public static native boolean lock(int key, long timeout);

    public static native void unlock(int key);

    public static native void waitForBufferChanges(int key);

    public static native boolean hasBufferChanges(int key);

    public static native void lockBuffer(int key);

    public static native void unlockBuffer(int key);
}

File: platform/src/main/java/com/toocol/termio/platform/nativefx/NativeBinding.kt
package com.toocol.termio.platform.nativefx

import javafx.scene.input.MouseEvent
import javafx.scene.input.ScrollEvent
import java.io.FileNotFoundException
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.util.*

class NativeBinding {
    internal interface IntEnum {
        fun asInt(): Int

        companion object {
            fun combine(vararg enums: IntEnum): Int {
                var result = 0
                for (e in enums) {
                    result = result or e.asInt()
                }
                return result
            }

            fun combine(combined: Int, vararg enums: IntEnum): Int {
                var result = combined
                for (e in enums) {
                    result = result or e.asInt()
                }
                return result
            }
        }
    }

    internal interface MouseBtn {
        companion object {
            const val NO_BTN = 0
            const val PRIMARY_BTN = 1
            const val SECONDARY_BTN = 2
            const val MIDDLE_BTN = 4

            fun fromEvent(ev: MouseEvent): Int {
                var result = NO_BTN
                if (ev.isPrimaryButtonDown) {
                    result = result or PRIMARY_BTN
                }
                if (ev.isSecondaryButtonDown) {
                    result = result or SECONDARY_BTN
                }
                if (ev.isMiddleButtonDown) {
                    result = result or MIDDLE_BTN
                }
                return result
            }

            fun fromEvent(ev: ScrollEvent?): Int {
                // TODO implement me 8.07.2019
                return NO_BTN
            }
        }
    }

    internal interface Modifier {
        companion object {
            const val NO_KEY = 0
            const val SHIFT_KEY = 1
            const val ALT_KEY = 2
            const val META_KEY = 4
            const val CONTROL_KEY = 8

            fun fromEvent(ev: MouseEvent): Int {
                var result = NO_KEY
                if (ev.isShiftDown) {
                    result = result or SHIFT_KEY
                }
                if (ev.isAltDown) {
                    result = result or ALT_KEY
                }
                if (ev.isMetaDown) {
                    result = result or META_KEY
                }
                if (ev.isControlDown) {
                    result = result or CONTROL_KEY
                }
                return result
            }

            fun fromEvent(ev: ScrollEvent): Int {
                var result = NO_KEY
                if (ev.isShiftDown) {
                    result = result or SHIFT_KEY
                }
                if (ev.isAltDown) {
                    result = result or ALT_KEY
                }
                if (ev.isMetaDown) {
                    result = result or META_KEY
                }
                if (ev.isControlDown) {
                    result = result or CONTROL_KEY
                }
                return result
            }
        }
    }

    init {
        throw AssertionError("Can't instantiated.")
    }

    companion object {
        private var initialized = false

        external fun nextKey(): Int
        external fun connectTo(name: String?): Int
        external fun terminate(key: Int): Boolean
        external fun isConnected(key: Int): Boolean
        external fun sendMsg(key: Int, msg: String?): String?

        external fun processNativeEvents(key: Int)
        external fun resize(key: Int, w: Int, h: Int)
        external fun isDirty(key: Int): Boolean
        external fun redraw(key: Int, x: Int, y: Int, w: Int, h: Int)
        external fun setDirty(key: Int, value: Boolean)
        external fun setBufferReady(key: Int, value: Boolean)
        external fun isBufferReady(key: Int): Boolean
        external fun getW(key: Int): Int
        external fun getH(key: Int): Int
        external fun fireMousePressedEvent(
            key: Int,
            x: Double,
            y: Double,
            buttons: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireMouseReleasedEvent(
            key: Int,
            x: Double,
            y: Double,
            buttons: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireMouseClickedEvent(
            key: Int,
            x: Double,
            y: Double,
            buttons: Int,
            modifiers: Int,
            clickCount: Int,
            timestamp: Long,
        ): Boolean

        external fun fireMouseEnteredEvent(
            key: Int,
            x: Double,
            y: Double,
            buttons: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireMouseExitedEvent(
            key: Int,
            x: Double,
            y: Double,
            buttons: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireMouseMoveEvent(
            key: Int,
            x: Double,
            y: Double,
            buttons: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireMouseWheelEvent(
            key: Int,
            x: Double,
            y: Double,
            amount: Double,
            buttons: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireKeyPressedEvent(
            key: Int,
            characters: String?,
            keyCode: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireKeyReleasedEvent(
            key: Int,
            characters: String?,
            keyCode: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun fireKeyTypedEvent(
            key: Int,
            characters: String?,
            keyCode: Int,
            modifiers: Int,
            timestamp: Long,
        ): Boolean

        external fun getBuffer(key: Int): ByteBuffer?
        external fun lock(key: Int): Boolean
        external fun lock(key: Int, timeout: Long): Boolean
        external fun unlock(key: Int)
        external fun waitForBufferChanges(key: Int)
        external fun hasBufferChanges(key: Int): Boolean
        external fun lockBuffer(key: Int)
        external fun unlockBuffer(key: Int)

        fun libEnding(): String {
            if (isOS("windows")) return ".dll"
            if (isOS("linux")) return ".so"
            return if (isOS("macos")) ".dylib" else ".so"
        }

        fun archName(): String {
            val osArch = System.getProperty("os.arch")
            if (osArch.lowercase(Locale.getDefault()).contains("x64")
                || osArch.lowercase(Locale.getDefault()).contains("amd64")
                || osArch.lowercase(Locale.getDefault()).contains("x86_64")
            ) {
                return "x64"
            } else if (osArch.lowercase(Locale.getDefault()).contains("x86")) {
                return "x86"
            } else if (osArch.lowercase(Locale.getDefault()).contains("aarch64")) {
                return "aarch64"
            }
            return osArch
        }

        fun osName(): String {
            val osName = System.getProperty("os.name")
            if (osName.lowercase(Locale.getDefault()).contains("windows")) {
                return "windows"
            } else if (osName.lowercase(Locale.getDefault()).contains("linux")) {
                return "linux"
            } else if (osName.lowercase(Locale.getDefault()).contains("mac")
                || osName.lowercase(Locale.getDefault()).contains("darwin")
            ) {
                return "macos"
            }
            return osName
        }

        fun isOS(os: String): Boolean {
            return os == osName()
        }

        fun init() {
            if (initialized) {
                // throw new RuntimeException("Already initialized.");
                return
            }
            initialized = true
            var path = "/eu/mihosoft/nativefx/nativelibs/"
            var vcredistPath1 = path
            var vcredistPath2 = path
            var libName = "nativefx" + libEnding()
            if (isOS("windows")) {
                path += "windows/" + archName() + "/" + libName
                vcredistPath1 += "windows/" + archName() + "/vcruntime140.dll"
                vcredistPath2 += "windows/" + archName() + "/msvcp140.dll"
            } else if (isOS("linux")) {
                libName = "lib$libName"
                path += "linux/" + archName() + "/" + libName
            } else if (isOS("macos")) {
                libName = "lib$libName"
                path += "macos/" + archName() + "/" + libName
            }
            try {
                val libPath = Files.createTempDirectory("nativefx-libs")
                if (isOS("windows")) {
                    resourceToFile(vcredistPath1, Paths.get(libPath.toFile().absolutePath, "vcruntime140.dll"), false)
                    resourceToFile(vcredistPath2, Paths.get(libPath.toFile().absolutePath, "msvcp140.dll"), false)
                    if (Paths.get(libPath.toFile().absolutePath + "/vcruntime140.dll").toFile().isFile) {
                        println("> loading " + libPath.toFile().absolutePath + "/vcruntime140.dll")
                        System.load(libPath.toFile().absolutePath + "/vcruntime140.dll")
                    }
                }
                resourceToFile(path, Paths.get(libPath.toFile().absolutePath, libName))
                println("> loading " + libPath.toFile().absolutePath + "/" + libName)
                System.load(libPath.toFile().absolutePath + "/" + libName)
            } catch (e: IOException) {
                e.printStackTrace(System.err)
            }
        }

        @Throws(IOException::class)
        private fun resourceToFile(resource: String, destination: Path, failOnError: Boolean = true) {
            try {
                NativeBinding::class.java.getResourceAsStream(resource).use { `is` ->
                    println("> unpacking resource '$resource' to file '$destination'")
                    Files.copy(`is`, destination,
                        StandardCopyOption.REPLACE_EXISTING)
                }
            } catch (e: NullPointerException) {
                if (failOnError) {
                    throw FileNotFoundException("Resource '$resource' was not found.")
                } else {
                    System.err.println("WARNING: Resource '$resource' was not found.")
                }
            }
        }

        private val listeners: MutableMap<Int, MutableList<NativeEventListener>> = HashMap()
        fun addEventListener(key: Int, l: NativeEventListener) {
            // create list if not present
            if (!listeners.containsKey(key)) {
                val list: MutableList<NativeEventListener> = ArrayList()
                listeners[key] = list
            }
            listeners[key]!!.add(l)
        }

        fun removeEventListener(key: Int, l: NativeEventListener) {
            // create list if not present
            if (!listeners.containsKey(key)) {
                val list: MutableList<NativeEventListener> = ArrayList()
                listeners[key] = list
            }
            listeners[key]!!.remove(l)
        }

        fun removeEventListeners(key: Int) {
            // create list if not present
            if (!listeners.containsKey(key)) {
                val list: MutableList<NativeEventListener> = ArrayList()
                listeners[key] = list
            }
            listeners[key]!!.clear()
        }

        /*CALLED FROM NATIVE*/
        fun fireNativeEvent(key: Int, type: String?, evt: String?) {
            // return early if not present
            if (!listeners.containsKey(key)) {
                return
            }
            val list: List<NativeEventListener> = listeners[key]!!
            for (l in list) {
                l.event(key, type, evt)
            }
        }
    }
}